tokio-threadpool/src/sender.rs

```diff
@@ -87,6 +96,90 @@ impl Sender {
         tokio_executor::Executor::spawn(&mut s, Box::new(future))
     }
 
+    /// Run a future to completion on the thread pool.
+    ///
+    /// This function will block the caller until the given future has completed,
+    /// so it should not be called from an asynchronous context.
+    ///
+    /// If `block_on` returns `Err`, then the future failed to be run. There
+    /// are two possible causes:
+    ///
+    /// * The thread pool is at capacity and is unable to run a new future.
+    ///   This is a temporary failure. At some point in the future, the thread
+    ///   pool might be able to run new futures.
+    /// * The thread pool is shutdown. This is a permanent failure indicating
+    ///   that the handle will never be able to run new futures.
+    ///
+    /// The status of the thread pool can be queried before calling `block_on`
+    /// using the `status` function (part of the `Executor` trait).
+    ///
+    /// # Panics
+    ///
+    /// This method will panic if the provided future panics.
+    pub fn block_on<F>(&self, mut future: F) -> Result<Result<F::Item, F::Error>, SpawnError>
+        where F: Future + Send,
+              F::Item: Send,
+              F::Error: Send {
+
+        struct BlockOnFuture<F> where F: Future + Send {
+            future: *mut F,
+            result: *mut thread::Result<Result<F::Item, F::Error>>,
+            finished: *const AtomicBool,
+            thread: *const Thread,
+        }
+
+        unsafe impl <F> Send for BlockOnFuture<F>
+            where F: Future + Send,
+                  F::Item: Send,
+                  F::Error: Send {};
+
+        impl <F> Future for BlockOnFuture<F> where F: Future + Send {
+            type Item = ();
+            type Error = ();
+            fn poll(&mut self) -> Poll<(), ()> {
+                unsafe {
+                    let result = panic::catch_unwind(panic::AssertUnwindSafe(|| {
+                        (&mut *self.future).poll()
+                    }));
+                    let result = match result {
+                        Ok(Ok(Async::Ready(item))) => Ok(Ok(item)),
+                        Ok(Ok(Async::NotReady)) => return Ok(Async::NotReady),
+                        Ok(Err(error)) => Ok(Err(error)),
+                        Err(panic) => Err(panic),
+                    };
+                    ptr::write(self.result, result);
+                    (&*self.finished).store(true, Release);
+                    (&*self.thread).unpark();
+                }
+                Ok(Async::Ready(()))
+            }
+        }
+
+        let mut result: thread::Result<Result<F::Item, F::Error>> = unsafe { mem::uninitialized() };
+        let finished = AtomicBool::new(false);
+        let thread = thread::current();
+
+        let future: Box<Future<Item=(), Error=()>> = Box::new(BlockOnFuture {
```

Comment:
Why does this need to be wrapped in a `Box`?

Reply:
[`Executor::spawn`](https://docs.rs/tokio-executor/0.1.2/tokio_executor/trait.Executor.html#tymethod.spawn) takes a `Box`.